pkg/federate/kubefed/cluster_test.go

```diff
@@ -0,0 +1,252 @@
+package kubefed
+
+import (
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+	fedv1 "sigs.k8s.io/kubefed/pkg/apis/core/v1beta1"
+)
+
+type added string
+type updated string
+type removed string
+
+type testClusterEventHandler struct {
+	id     string
+	events chan interface{}
+}
+
+const (
+	secretName = "my-secret"
+)
+
+var _ = Describe("Kubefed Cluster", func() {
+	klog.InitFlags(nil)
+
+	Describe("Adding a watch", testAddingWatch)
+	Describe("When KubeFedCluster items are added, updated, removed", testOnKubeFedClusterChanges)
+	Describe("clusterWatcher", testClusterWatcher)
+})
+
+func testAddingWatch() {
+	var stopChan chan struct{}
+	BeforeEach(func() {
+		stopChan = make(chan struct{})
+	})
+
+	AfterEach(func() {
+		close(stopChan)
+	})
+
+	It("Should notify ClusterEventHandler of existing KubeFedCluster items", func() {
+		federator := newFederator(newInformer(watch.NewFake(), *newKubeFedCluster("east"),
+			*newKubeFedCluster("west")), stopChan)
+
+		handler1 := newtTestClusterEventHandler("1")
+		err := federator.WatchClusters(handler1)
+		Expect(err).ToNot(HaveOccurred())
+
+		Eventually(handler1.events, 5).Should(Receive(Or(Equal(added("east")), Equal(added("west")))))
+		Eventually(handler1.events, 5).Should(Receive(Or(Equal(added("east")), Equal(added("west")))))
```

Comment:
I love it! :+1:  amazing work tom!

Reply:
> I love it! 👍 amazing work tom!

Thanks!  The gomega lib is pretty good - `Eventually` and `Consistently` are really nice